<div style="border:solid black 2px; padding: 20px">

# Reviewer's Introduction

Halo Jeffry!

Saya Yosef, pada kesempatan kali ini saya akan memeriksa proyek kamu.

Kamu bisa melihat komentar saya di bagian bawah kode - tolong **jangan dipindah, diubah, atau dihapus** ya supaya saya bisa cek kembali ketika sudah kamu perbaiki.
    
Komentar dari saya bisa dalam kotak berwarna hijau, kuning, atau merah sebagai berikut

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Ada catatan.
</div>

<div class="alert alert-block alert-danger">

<b>Reviewer's comment</b> <a class="tocSkip"></a>

Perlu perbaikan.
</div>

Jeffry bisa merespon komentar saya dengan kotak berwarna biru, seperti ini

<div class="alert alert-block alert-info">
<b>Student's comment</b> <a class="tocSkip"></a>
    
Baik kak. <a class="tocSkip"></a>
</div>
    
Mohon diingat beberapa hal penting terkait proyek ini:
- proyek belum bisa diterima jika masih ada komentar dengan kotak berwarna merah
- tenang saja, kamu punya 3 kali kesempatan memperbaiki
- tapi ingat, pada setiap kesempatan kamu punya waktu maksimal 1 x 24 jam untuk memperbaiki proyek ya
    
</div>

<div style="border:solid black 1px; padding: 20px" class="alert alert-block alert-warning">
    
# Reviewer's general feedback v1

Terima kasih Jeffry sudah mengirimkan proyek ini!

Secara umum proyek kamu sudah **sangat baik**! Hanya ada beberapa catatan dalam kotak berwarna kuning dan merah yang bisa kamu perbaiki dengan mudah, setelah itu proyek ini bisa diterima. Jangan lupa untuk menghapus kalimat-kalimat instruksi yang ditandakan dengan simbol `[..]` dan untuk deskripsi/jawaban yang kamu tulis sendiri bisa kamu ubah menjadi cell markdown tersendiri. Ibaratkan notebook ini akan dibaca oleh orang lain, bisa jadi tim bisnis, yang perlu kamu bantu dalam memahami setiap langkah yang kamu lakukan pada proyek ini dan kesimpulan apa saja yang bisa diambil sebagai insight. Untuk cell yang berisi kode yang sudah tidak dijalankan (kode yang sudah kamu buat sebagai komentar) juga bisa dihapus agar notebook semakin efisien.
    
Jika kamu memiliki pertanyaan, kamu bisa tuliskan dalam kotak berwarna biru seperti sudah dicontohkan di atas ya.
    
Semangat selalu Jeffry!

</div>

<div style="border:solid black 1px; padding: 20px" class="alert alert-block alert-success">
    
# Reviewer's general feedback v2

Terima kasih Jeffry sudah memperbaiki proyek ini!

Proyek kamu sudah bagus dan bisa diterima, tetapi jangan lupa untuk menghapus kalimat-kalimat instruksi yang ditandakan dengan simbol `[..]` ya dan untuk cell yang hanya berisi kode dalam bentuk komentar (kode yang sudah tidak digunakan lagi) bisa kamu hapus supaya notebook berisi kode yang memang dieksekusi saja.
  
*Good luck* dalam perjalanan belajarmu ya Jeffry!

</div>

# Menganalisis Risiko Gagal Bayar Peminjam

Tugas Anda adalah menyiapkan laporan untuk divisi kredit suatu bank. Anda akan mencari tahu pengaruh status perkawinan seorang nasabah dan jumlah anak yang dimilikinya terhadap probabilitas gagal bayar dalam pelunasan pinjaman. Pihak bank sudah memiliki beberapa data mengenai kelayakan kredit nasabah.

Laporan Anda akan dipertimbangkan pada saat membuat **penilaian kredit** untuk calon nasabah. **Penilaian kredit** digunakan untuk mengevaluasi kemampuan calon peminjam untuk melunasi pinjaman mereka.

[Dalam *notebook* ini, Anda dapat menemukan petunjuk dan instruksi singkat serta arahan untuk berfikir. Jangan abaikan petunjuk tersebut karena mereka dirancang untuk membekali Anda dengan struktur pengerjaan proyek dan akan membantu Anda melakukan analisis secara lebih mendalam. Sebaliknya, buatlah agar laporan ini terlihat seolah-olah dikirimkan oleh Anda kepada rekan satu tim Anda untuk menunjukkan temuan Anda - mereka tidak perlu tahu bahwa Anda mendapatkan bantuan eksternal dari kami! Untuk membantu Anda, kami menempatkan semua petunjuk yang harus Anda hapus dalam tanda kurung siku.]

[Sebelum Anda masuk ke dalam proses analisis data Anda, jelaskan tujuan proyek dan hipotesis yang akan Anda uji.]

Tujuan utama dari poject ini adalah untuk mengetahui kelayakan seorang klien untuk mendapatkan kredit berdasarkan status dan keadaan mereka yang tersimpan dalam data kita. Kita juga menguji kapasitas nasabah berdasarkan karakteristik mereka yang kita rangkum berdasarkan kategori-kategori sehingga diperoleh pattern untuk memberikan lampu kuning kepada nasabah yang masuk ke dalam kategori tertentu.

Hipotesis project :
1. Apakah terdapat korelasi antara jumlah anak dengan kemampuan melunasi pinjaman tepat waktu?
2. Apakah terdapat korelasi antara status keluarga dengan kemampuan melunasi pinjaman tepat waktu?
3. Apakah terdapat korelasi antara kelas ekonomi dengan kemampuan melunasi pinjaman tepat waktu?
4. Apakah terdapat korelasi antara tujuan kredit dengan kemampuan melunasi pinjaman tepat waktu?

<div class="alert alert-block alert-danger">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Jangan lupa tuliskan deskripsi, tujuan, dan hipotesis proyek pada setiap awal notebook ya. Kamu bisa melihat contohnya seperti pada proyek sebelumnya.
</div>

<div class="alert alert-block alert-success">

<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Sip, bagus sekali.
</div>

## Buka *file* data dan baca informasi umumnya.

[Mulailah dengan mengimpor *library* dan memuat datanya. Anda mungkin akan menyadari bahwa Anda memerlukan *library* tambahan saat Anda telah mengerjakan proyek ini, dan itu merupakan hal yang normal. Hanya saja, pastikan untuk memperbarui bagian ini jika dibutuhkan.]

In [1]:
import pandas as pd # Muat semua *library*


df=pd.read_csv('/datasets/credit_scoring_eng.csv')# Muat datanya


<div class="alert alert-block alert-warning">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Oke sudah bagus, tetapi lebih baik antara pemuatan library dan data dipisah menjadi 2 cell supaya ketika nantinya kamu perlu menambahkan library baru, kamu tidak perlu memuat ulang seluruh datanya.
</div>

In [2]:
df

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 12 columns]

## Soal 1. Eksplorasi data

**Deskripsi Data**
- `children` - jumlah anak dalam keluarga
- `days_employed` - pengalaman kerja nasabah dalam hari
- `dob_years` - usia nasabah dalam tahun
- `education` - tingkat pendidikan nasabah
- `education_id` - pengidentifikasi untuk tingkat pendidikan nasabah
- `family_status` - pengidentifikasi untuk status perkawinan nasabah
- `family_status_id` - tanda pengenal status perkawinan
- `gender` - jenis kelamin nasabah
- `income_type` - jenis pekerjaan
- `debt` - apakah nasabah pernah melakukan gagal bayar pinjaman
- `total_income` - pendapatan bulanan
- `purpose` - tujuan mendapatkan pinjaman

[Sekarang saatnya menjelajahi data kita. Anda perlu melihat berapa banyak kolom dan baris yang dimiliki oleh data, serta mencermati beberapa baris data untuk memeriksa potensi masalah dengan data.]

In [3]:
df.shape# Mari kita lihat berapa banyak baris dan kolom yang dimiliki oleh dataset kita



(21525, 12)

In [4]:
df.head()# Mari tampilkan N baris pertama



children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

[Jelaskan apa yang Anda lihat dan perhatikan dalam sampel data yang Anda tampilkan. Apakah terdapat masalah yang mungkin memerlukan penyelidikan dan perubahan lebih lanjut?]

In [5]:
df.info()# Dapatkan informasi data


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [6]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

[Apakah terdapat nilai yang hilang di semua kolom atau hanya di beberapa kolom? Jelaskan secara singkat apa yang Anda amati dalam 1-2 kalimat.]
Kolom'days_employed' dan 'total_income' sama-sama memiliki data yang hilang sebanyak 2174 dari total 21525. 


<div class="alert alert-block alert-warning">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Tambahkan juga pengecekan menggunakan `describe()` untuk mendapat gambaran karakteristik kolom numerik ya.
</div>

<div class="alert alert-block alert-success">

<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Oke, sudah ditambahkan `describe()`.
</div>

In [7]:
# Mari kita lihat tabel yang telah difilter dengan nilai yang hilang di kolom pertama yang mengandung data yang hilang
df[df['days_employed'].isnull()]


children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

[Apakah nilai yang hilang tampak simetris? Apakah kita yakin dengan asumsi ini? Jelaskan pendapat Anda secara singkat di bagian ini. Anda mungkin ingin melakukan penyelidikan lebih lanjut, dan menghitung nilai yang hilang di semua baris dengan nilai yang hilang untuk memastikan bahwa sampel yang hilang memiliki ukuran yang sama.]
Jika hanya melihat jumlah baris yang hilang yaitu 2174 apabila dijumlahkan dengan 19351 maka hasilnya 21525, artinya simetris. Tapi ada beberapa data di kolom 'days_employed' yang nilainya aneh (minus dan besar), maka perlu ditelaah lebih lanjut. 

In [8]:
# Mari kita terapkan beberapa kondisi untuk memfilter data dan melihat jumlah baris dalam tabel yang telah difilter.
df.loc[(df['days_employed'].isna()) & (df['total_income'].isna())].shape


(2174, 12)

**Kesimpulan sementara**

[Apakah jumlah baris dalam tabel yang telah difilter sesuai dengan jumlah nilai yang hilang? Kesimpulan apa yang bisa kita buat dari hal ini?]

[Hitung persentase nilai yang hilang jika dibandingkan dengan keseluruhan *dataset*. Apakah nilai yang hilang merupakan bagian data yang cukup besar? Jika demikian, Anda sebaiknya perlu mengisi nilai yang hilang. Untuk melakukannya, pertama-tama kita harus mempertimbangkan apakah data yang hilang bisa jadi disebabkan oleh karakteristik nasabah tertentu, seperti jenis pekerjaan atau yang lainnya. Anda harus memutuskan karakteristik mana yang menurut *Anda* mungkin merupakan penyebabnya. Kedua, kita harus memeriksa apakah ada ketergantungan nilai yang hilang pada nilai indikator lain dengan kolom-kolom yang mengidentifikasikan karakteristik tertentu nasabah.]

[Jelaskan langkah Anda selanjutnya dan bagaimana hubungannya dengan kesimpulan yang Anda buat sejauh ini.]

In [9]:
# Mari kita periksa nasabah yang tidak memiliki data tentang karakteristik yang teridentifikasi dan kolom dengan nilai yang hilang

df_null = df[(df['days_employed'].isnull()) & (df['total_income'].isnull())]
df_null.head(20)

children  days_employed  dob_years            education  education_id  \
12          0            NaN         65  secondary education             1   
26          0            NaN         41  secondary education             1   
29          0            NaN         63  secondary education             1   
41          0            NaN         50  secondary education             1   
55          0            NaN         54  secondary education             1   
65          0            NaN         21  secondary education             1   
67          0            NaN         52    bachelor's degree             0   
72          1            NaN         32    bachelor's degree             0   
82          2            NaN         50    bachelor's degree             0   
83          0            NaN         52  secondary education             1   
90          2            NaN         35    bachelor's degree             0   
94          1            NaN         34    bachelor's degree             0   
96          0            NaN         44  SECONDARY EDUCATION             1   
97          0            NaN         47    bachelor's degree             0   
120         0            NaN         46  secondary education             1   
121         0            NaN         29    bachelor's degree             0   
135         0            NaN         27  secondary education             1   
141         0            NaN         39  secondary education             1   
145         0            NaN         62  secondary education             1   
174         0            NaN         55    bachelor's degree             0   

         family_status  family_status_id gender    income_type  debt  \
12   civil partnership                 1      M        retiree     0   
26             married                 0      M  civil servant     0   
29           unmarried                 4      F        retiree     0   
41             married                 0      F  civil servant     0   
55   civil partnership                 1      F        retiree     1   
65           unmarried                 4      M       business     0   
67             married                 0      F        retiree     0   
72             married                 0      M  civil servant     0   
82             married                 0      F       employee     0   
83             married                 0      M       employee     0   
90             married                 0      F       employee     0   
94   civil partnership                 1      F       business     0   
96             married                 0      F       employee     0   
97             married                 0      F       employee     0   
120            married                 0      F       employee     0   
121            married                 0      F       business     0   
135            married                 0      M       business     0   
141  civil partnership                 1      M       employee     0   
145            married                 0      M        retiree     0   
174    widow / widower                 2      F       business     0   

     total_income                                   purpose  
12            NaN                         to have a wedding  
26            NaN                                 education  
29            NaN                    building a real estate  
41            NaN                  second-hand car purchase  
55            NaN                         to have a wedding  
65            NaN  transactions with commercial real estate  
67            NaN       purchase of the house for my family  
72            NaN  transactions with commercial real estate  
82            NaN                                   housing  
83            NaN                                   housing  
90            NaN                      housing transactions  
94            NaN                          having a wedding  
96            NaN               buy residential

In [10]:
# Periksalah distribusinya
df_null = df_null.shape[0] / df.shape[0] 
print(f'Distribusi nilai yang hilang sebesar: {df_null:0%}')

Distribusi nilai yang hilang sebesar: 10.099884%


In [11]:
df_null = df[(df['days_employed'].isnull()) & (df['total_income'].isnull())]

[Deksripsikan temuan Anda di sini.]
**Kemungkinan penyebab hilangnya nilai dalam data**
melihat dari persebaran data, dari total 2174 rows atau sekitar 10% dari total 21252 rows, dimana nilai hilang tidak hanya pada income_type atau purpose maupun family_status tertentu saja, tapi tersebar. Saya melihat hilangnya nilai data ini tampaknya acak. 

[Kemukakan gagasan Anda mengenai kemungkinan alasan hilangnya nilai-nilai tersebut. Apakah menurut Anda mereka hilang secara acak atau apakah terdapat pola-pola tertentu?]

[Mari kita mulai memeriksa apakah nilai yang hilang bersifat acak.]

<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Oke, analisis nilai hilang sudah bagus.
</div>

In [12]:
# Memeriksa distribusi di seluruh *dataset*

df.isna().sum() / df.shape[0] * 100

children             0.000000
days_employed       10.099884
dob_years            0.000000
education            0.000000
education_id         0.000000
family_status        0.000000
family_status_id     0.000000
gender               0.000000
income_type          0.000000
debt                 0.000000
total_income        10.099884
purpose              0.000000
dtype: float64

**Kesimpulan sementara**

[Apakah distribusi dalam *dataset* yang asli mirip dengan distribusi tabel yang telah difilter? Apa artinya hal tersebut untuk kita?] Ada kemiripan antara dataset asli dan distribusi tabel filtered. Artinya missing values bersifat random, tidak bergantung pada 1 atau 2 kolom saja. 

[Jika menurut Anda kita belum dapat membuat kesimpulan apa pun, mari kita kembali menyelidiki *dataset* lebih lanjut. Mari pikirkan alasan lain yang dapat menyebabkan data hilang dan periksa apakah kita dapat menemukan pola tertentu yang dapat membuat kita berpikir bahwa hilangnya nilai-nilai tersebut tidak terjadi secara acak. Karena ini merupakan tugas Anda, maka bagian ini adalah opsional.]

In [13]:
# Periksa penyebab dan pola lain yang dapat mengakibatkan nilai yang hilang
df_null['income_type'].value_counts(normalize=True)*100

employee         50.827967
business         23.367065
retiree          18.997240
civil servant     6.761730
entrepreneur      0.045998
Name: income_type, dtype: float64

In [14]:
df_null['purpose'].value_counts(normalize=True)*100

having a wedding                            4.231831
to have a wedding                           3.725851
wedding ceremony                            3.495860
construction of own property                3.449862
housing transactions                        3.403864
buy real estate                             3.311868
transactions with my real estate            3.265869
purchase of the house for my family         3.265869
housing renovation                          3.219871
transactions with commercial real estate    3.219871
buy commercial real estate                  3.081877
buying property for renting out             2.989880
property                                    2.851886
real estate transactions                    2.805888
buy residential real estate                 2.805888
housing                                     2.759890
building a property                         2.713891
cars                                        2.621895
going to university                         2.

In [15]:
df_null['family_status'].value_counts(normalize=True)*100

married              56.899724
civil partnership    20.331187
unmarried            13.247470
divorced              5.151794
widow / widower       4.369825
Name: family_status, dtype: float64

In [16]:
df_null['children'].value_counts(normalize=True)*100

 0     66.191352
 1     21.849126
 2      9.383625
 3      1.655934
 20     0.413983
 4      0.321987
-1      0.137994
 5      0.045998
Name: children, dtype: float64

melihat dari beberapa kolom dimana terdapat nilai hilang pada kolom days_employed dan total_income, justru yang 
paling banyak nilai hilang adalah kelompok produktif TAPI tidak punya anak atau hanya 1 anak.
Purpose orang yang mengajukan kredit didominasi oleh issues wedding dan property. 
Dengan melihat hal-hal diatas sepertinya terpola, tapi dasar dari nilai yang hilang pada kolom days_employed
dan total_income belum tahu pasti. 
Dugaan sementara sy: perihal days_employed dan total_income merupakan data sensitif yang kebanyakan orang enggan
untuk membukanya bahkan kepada pihak pemberi kredit. Adanya manipulasi data untuk tercapainya tujuan juga sering 
menjadi alasan untuk pihak pemberi kredit memilih untuk menganalisa rekening koran pada account giro atau tabungan
calon debitur. 

<div class="alert alert-block alert-warning">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Analisis kamu pada cell di atas sudah sangat bagus, bisa kamu ubah cell di atas menjadi tipe markdown saja supaya lebih rapih ya. Semua deskripsi dan penjelasan yang kamu tulis pada notebook ini sebaiknya kamu tulis dalam cell markdown.
</div>

<div class="alert alert-block alert-success">

<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Sip.
</div>

In [17]:
#df.loc[df['days_employed'].isna(),'education'].str.lower().unique()

In [18]:
#df['education'].str.lower().unique()

In [19]:
#df['education'].str.lower().value_counts(normalize=True)

In [20]:
#df['purpose'].value_counts(normalize=True).reset_index()

In [21]:
#df.loc[df['total_income'].isna(),'purpose'].unique().shape

**Kesimpulan sementara**

[Apakah pada akhirnya kita dapat memastikan bahwa nilai yang hilang adalah suatu kebetulan? Periksa hal lain yang menurut Anda penting di sini.]

In [22]:
# Periksa pola lainnya - jelaskan pola tersebut
df_null['gender'].value_counts(normalize=True)*100

F    68.26127
M    31.73873
Name: gender, dtype: float64

**Kesimpulan**

[Apakah Anda menemukan suatu pola? Bagaimana Anda mendapatkan kesimpulan ini?]
Nilai yang hilang pada kolom days_employed dan total_income dapat digambarkan sbb.:
Pemohon kredit didominasi oleh female (hampir 70%) dengan status married atau civil partnership dimana memiliki pekerjaan terutama sebagai employee, dengan tujuan terutama untuk wedding (mungkin bagi civil partnership) dan property(mungkin bagi yang telah married maupun yang civil partnership). 

[Jelaskan bagaimana Anda akan mengatasi nilai-nilai yang hilang. Pertimbangkan kategori dengan nilai yang hilang.]

[Rencanakan secara singkat langkah Anda selanjutnya untuk mentransformasi data. Anda mungkin perlu mengatasi berbagai jenis masalah: duplikat, pencatatan yang berbeda, sumber data yang salah, dan nilai yang hilang.]

Hasil analisis sampai saat ini menunjukkan nilai yang hilang bersifat acak dan tidak bergantung pada kolom tertentu

<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Oke, bisa kamu tuliskan lagi pada kesimpulan sementara bahwa hasil analisis menunjukkan nilai yang hilang bersifat acak dan tidak bergantung pada kolom tertentu.
</div>

## Transformasi data

[Mari kita perhatikan setiap kolom untuk melihat masalah apa yang mungkin dimiliki mereka.]

[Mulailah dengan menghapus duplikat dan memperbaiki data tentang informasi pendidikan jika diperlukan.]

In [23]:
# Mari kita lihat semua nilai di kolom pendidikan untuk memeriksa ejaan apa yang perlu diperbaiki
df['education'].unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

In [24]:
# Perbaiki pencatatan jika diperlukan

df['education']=df['education'].str.lower()

<div class="alert alert-block alert-danger">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Menggunakan `.str.lower()` sudah tepat, tidak perlu menggunakan `.head()` ya.
</div>

<div class="alert alert-block alert-success">

<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Oke, sudah bagus.
</div>

[Periksa data kolom `children`]

In [26]:
# Mari kita lihat distribusi nilai pada kolom `children`
df['children'].value_counts(normalize=True)*100

 0     65.732869
 1     22.383275
 2      9.547038
 3      1.533101
 20     0.353078
-1      0.218351
 4      0.190476
 5      0.041812
Name: children, dtype: float64

In [27]:
df['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

In [28]:
df[(df['children']==-1)|(df['children']==20)][['children','income_type','purpose']].value_counts()
    

children  income_type  purpose                                 
 20       employee     buying a second-hand car                    4
                       buy commercial real estate                  3
                       transactions with commercial real estate    3
                       to own a car                                3
          business     university education                        3
                                                                  ..
-1        employee     to have a wedding                           1
                       supplementary education                     1
                       second-hand car purchase                    1
                       real estate transactions                    1
 20       retiree      transactions with commercial real estate    1
Length: 94, dtype: int64

In [29]:
#df.loc[(df['children']==-1)|(df['children']==20),'children']=df['children'].median()

In [30]:
#df['children'].value_counts(normalize=True)

[Apakah terdapat hal-hal yang aneh di kolom tersebut? Bagaimana mereka bisa terjadi? Buat keputusan tentang apa yang akan Anda lakukan dengan data ini dan jelaskan alasannya.]

In [31]:
# [perbaiki data berdasarkan keputusan Anda]
df.loc[(df['children']==-1)|(df['children']==20),'children']=df['children'].median()

Tidak ada data yang mendukung untuk menghapus jumlah anak -1 dan 20 WALAUPUN itu jelas buat saya sebagai outlier
Sehingga mengganti nilai -1 dan 20 menjadi median adalah keputusan yang moderat namun tidak merubah distribusi 
data secara signifikan. 

<div class="alert alert-block alert-warning">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Kamu perlu jelaskan kenapa median adalah nilai yang tepat untuk mengisi angka aneh pada kolom `children`.
</div>

<div class="alert alert-block alert-success">

<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Oke, mengganti dengan median bisa diterima. Asumsi lain yang bisa kamu pertimbangkan adalah mengganti -1 dengan 1 dan 20 dengan 2.
</div>

In [32]:
df['children'].value_counts(normalize=True)*100

0.0    66.304297
1.0    22.383275
2.0     9.547038
3.0     1.533101
4.0     0.190476
5.0     0.041812
Name: children, dtype: float64

<div class="alert alert-block alert-warning">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Ubah cell di atas menjadi tipe markdown ya.
</div>

<div class="alert alert-block alert-success">

<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Sip.
</div>

In [33]:
#df = df[(df.children != -1) & (df.children != 20)]
#df['children'].value_counts()

In [34]:
# Periksa kembali kolom `children` untuk memastikan bahwa semuanya telah diperbaiki

#df['children'].value_counts(normalize=True)

[Periksa data dalam kolom the `days_employed`. Pertama-tama pikirkan tentang masalah apa yang mungkin ada pada kolom tersebut, serta pikirkan juga apa yang mungkin ingin Anda periksa dan bagaimana Anda akan melakukannya.]

In [35]:
# Temukan data yang bermasalah di kolom `days_employed` jika memang terdapat masalah, dan hitung persentasenya
df['days_employed']

0         -8437.673028
1         -4024.803754
2         -5623.422610
3         -4124.747207
4        340266.072047
             ...      
21520     -4529.316663
21521    343937.404131
21522     -2113.346888
21523     -3112.481705
21524     -1984.507589
Name: days_employed, Length: 21525, dtype: float64

In [36]:
df['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

In [37]:
df[df['days_employed']<0].shape

(15906, 12)

In [38]:
df[(df['days_employed']<0)]['income_type'].value_counts(normalize=True)*100

employee                       62.957375
business                       28.775305
civil servant                   8.248460
entrepreneur                    0.006287
paternity / maternity leave     0.006287
student                         0.006287
Name: income_type, dtype: float64

In [39]:
df[(df['days_employed']>0)]['income_type'].value_counts(normalize=True)*100

retiree       99.941945
unemployed     0.058055
Name: income_type, dtype: float64

In [40]:
wrong_data_days = df[df['days_employed']<0]
wrong_data_days['days_employed'].reset_index(drop=True)

0       -8437.673028
1       -4024.803754
2       -5623.422610
3       -4124.747207
4        -926.185831
            ...     
15901   -2351.431934
15902   -4529.316663
15903   -2113.346888
15904   -3112.481705
15905   -1984.507589
Name: days_employed, Length: 15906, dtype: float64

In [41]:
percentage_days = wrong_data_days['days_employed'].count() / df['days_employed'].count()
print("Persentase data yang bermasalah pada kolom 'days_employed' yaitu:{:.0%} dari keseluruhan data.".format(percentage_days))

Persentase data yang bermasalah pada kolom 'days_employed' yaitu:82% dari keseluruhan data.


[Jika jumlah data yang bermasalah tinggi, hal tersebut mungkin disebabkan oleh beberapa masalah teknis. Kita mungkin perlu mengusulkan alasan paling jelas mengapa hal tersebut dapat terjadi dan bagaimana seharusnya data yang benar, mengingat kita tidak dapat menghapus baris yang bermasalah ini.]
kemungkinan nilai pada days_employed merupakan kode dari credit scoring dimana nilai minus diberikan kepada kelompok produktif seperti employee dan business, sedangkan nilai positif justru diberikan pada retiree dan
unemployed. Tapi untuk memudahkan analisa dalam konteks ini, baiklah semua nilai pada kolom days_employed ini diubah menjadi positif. 

In [42]:
# Atasi nilai yang bermasalah, jika ada

df['days_employed'] = df['days_employed'].abs()

In [43]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
df['days_employed'].describe()

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Oke, menggunakan `.abs()` sudah tepat untuk `days_employed`.
</div>

In [44]:
#min(df['days_employed'])

[Sekarang mari kita lihat usia nasabah dan mengecek apakah terdapat masalah di sana. Sekali lagi, pikirkan tentang kemungkinan kejanggalan apa yang bisa kita temui dalam kolom ini, misalnya angka usia yang tidak masuk akal.]

In [45]:
# Periksa `dob_years` untuk nilai yang mencurigakan dan hitung persentasenya

df['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [46]:
df['dob_years'].median()

42.0

In [47]:
df['dob_years'].value_counts(normalize=True)*100

35    2.866434
40    2.829268
41    2.819977
34    2.801394
38    2.778165
42    2.773519
33    2.699187
39    2.662021
31    2.601626
36    2.578397
44    2.541231
29    2.531940
30    2.508711
48    2.499419
37    2.494774
50    2.387921
43    2.383275
32    2.369338
49    2.360046
28    2.336818
45    2.308943
27    2.290360
56    2.262485
52    2.248548
47    2.229965
54    2.225319
46    2.206736
58    2.141696
57    2.137050
53    2.132404
51    2.081301
59    2.062718
55    2.058072
26    1.895470
60    1.751452
25    1.658537
61    1.649245
62    1.635308
63    1.249710
64    1.231127
24    1.226481
23    1.180023
65    0.901278
66    0.850174
22    0.850174
67    0.775842
21    0.515679
0     0.469222
68    0.459930
69    0.394890
70    0.301974
71    0.269454
20    0.236934
72    0.153310
19    0.065041
73    0.037166
74    0.027875
75    0.004646
Name: dob_years, dtype: float64

In [48]:
df.loc[df['dob_years']==0].shape

(101, 12)

In [49]:
df.loc[df['dob_years']==0]['income_type'].value_counts()

employee         55
retiree          20
business         20
civil servant     6
Name: income_type, dtype: int64

In [50]:
df.loc[df['dob_years']==0]['family_status'].value_counts()

married              49
civil partnership    21
unmarried            16
divorced             10
widow / widower       5
Name: family_status, dtype: int64

In [51]:
pd.DataFrame(df['dob_years'].value_counts(normalize=True).sort_index()).T

0        19        20        21        22      23        24  \
dob_years  0.004692  0.00065  0.002369  0.005157  0.008502  0.0118  0.012265   

                 25        26        27  ...        66        67        68  \
dob_years  0.016585  0.018955  0.022904  ...  0.008502  0.007758  0.004599   

                 69       70        71        72        73        74        75  
dob_years  0.003949  0.00302  0.002695  0.001533  0.000372  0.000279  0.000046  

[1 rows x 58 columns]

[Putuskan apa yang akan Anda lakukan dengan nilai yang bermasalah dan jelaskan alasannya.]
Saya akan mengganti nilai yang hilang dengan median dari kolom dob_years

In [52]:
# Atasi masalah pada kolom `dob_years`, jika terdapat masalah
df_days_employed_prob_percent = (df['dob_years'] == 0).sum() / df.shape[0]
print(f'Persentase data yang bermasalah adalah: {df_days_employed_prob_percent:.2%}')

Persentase data yang bermasalah adalah: 0.47%


In [53]:
df.loc[(df['dob_years']==0),'dob_years']=df['dob_years'].median()

Dengan distribusi data rata-rata dob_years merata di 0-2% (lihat value_counts) dan data yang bermasalah 'hanya' <0.5% maka menurut saya penempatan nilai median pada data-data yang bermasalah tidak akan merusak informasi data yang utuh (secara keseluruhan) dan dapat dipertanggungjawabkan. 

In [54]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
df.loc[(df['dob_years'].isna())].shape

(0, 12)

<div class="alert alert-block alert-warning">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Jelaskan juga ya kenapa menggunakan `median` tepat untuk mengisi angka aneh pada `dob_years`.
</div>

<div class="alert alert-block alert-success">

<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Sip.
</div>

[Sekarang saatnya memeriksa kolom `family_status`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu Anda atasi.]

In [55]:
# Mari kita lihat nilai untuk kolom ini
df['family_status'].value_counts(normalize=True)*100


married              57.514518
civil partnership    19.405343
unmarried            13.068525
divorced              5.551684
widow / widower       4.459930
Name: family_status, dtype: float64

In [56]:
df['family_status'].describe()

count       21525
unique          5
top       married
freq        12380
Name: family_status, dtype: object

In [57]:
df.loc[(df['family_status'].isna())].shape

(0, 12)

In [58]:
# Atasi nilai yang bermasalah di `family_status`, jika ada

#tidak ada masalah 

In [59]:
# Periksa hasilnya - pastikan nilainya telah diperbaiki


[Sekarang saatnya memeriksa kolom `gender`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu Anda atasi]

In [60]:
# Mari kita liat nilai dalam kolom ini
df['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [61]:
df['gender'].value_counts(normalize=True)*100

F      66.137050
M      33.858304
XNA     0.004646
Name: gender, dtype: float64

In [62]:
# Atasi nilai-nilai yang bermasalah, jika ada
# gender XNA di drop saja karena distribusinya kurang dari 0.5% dan banyak NaN 
df_gender_value = df.pivot_table(index='gender', values='family_status', columns='purpose', aggfunc= 'count')
df_gender_value

purpose  building a property  building a real estate  \
gender                                                 
F                      412.0                   417.0   
M                      208.0                   209.0   
XNA                      NaN                     NaN   

purpose  buy commercial real estate  buy real estate  \
gender                                                 
F                             447.0            395.0   
M                             217.0            228.0   
XNA                             NaN              1.0   

purpose  buy residential real estate  buying a second-hand car  \
gender                                                           
F                              407.0                     324.0   
M                              200.0                     155.0   
XNA                              NaN                       NaN   

purpose  buying my own car  buying property for renting out    car  \
gender                                                               
F                    338.0                            443.0  339.0   
M                    167.0                            210.0  156.0   
XNA                    NaN                              NaN    NaN   

purpose  car purchase  ...  supplementary education  to become educated  \
gender                 ...                                                
F               320.0  ...                    311.0               266.0   
M               142.0  ...                    151.0               146.0   
XNA               NaN  ...                      NaN                 NaN   

purpose  to buy a car  to get a supplementary education  to have a wedding  \
gender                                                                       
F               295.0                             304.0              506.0   
M               177.0                             143.0              268.0   
XNA               NaN                               NaN                NaN   

purpose  to own a car  transactions with commercial real estate  \
gender                                                            
F               313.0                                     427.0   
M               167.0                                     224.0   
XNA               NaN                                       NaN   

purpose  transactions with my real estate  university education  \
gender                                                            
F                                   408.0                 298.0   
M                                   222.0                 155.0   
XNA                                   NaN                   NaN   

purpose  wedding ceremony  
gender                     
F                   547.0  
M                   250.0  
XNA                   NaN  

[3 rows x 38 columns]

In [63]:
df = df[(df.gender != 'XNA')]
df['gender'].value_counts()

F    14236
M     7288
Name: gender, dtype: int64

<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Sip, untuk `family_status` dan `gender` sudah bagus.
</div>

In [64]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki



[Sekarang saatnya memeriksa kolom `income_type`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu Anda atasi]

In [65]:
# Mari kita lihat nilai dalam kolom ini
df['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

In [66]:
df['income_type'].value_counts(normalize=True)*100

employee                       51.658614
business                       23.620145
retiree                        17.914886
civil servant                   6.778480
unemployed                      0.009292
entrepreneur                    0.009292
paternity / maternity leave     0.004646
student                         0.004646
Name: income_type, dtype: float64

In [67]:
# Atasi nilai yang bermasalah, jika ada
# tidak ada masalah pada kolom income_type

In [68]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki



[Sekarang saatnya melihat apakah ada duplikat di dalam data kita. Jika kita menemukannya, Anda harus memutuskan apa yang akan Anda lakukan dengan duplikat tersebut dan menjelaskan alasannya.]

In [69]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [70]:
# Periksa duplikat
print(df.duplicated().sum())

df_duplicated_percent = df.duplicated().sum() / df.shape[0] 
print(f'Distribusi nilai duplikat sebesar: {df_duplicated_percent:.2%}')

72
Distribusi nilai duplikat sebesar: 0.33%


In [71]:
# Atasi duplikat, jika ada
df = df.drop_duplicates().reset_index(drop=True)


In [72]:
# Lakukan pemeriksaan terakhir untuk mengecek apakah kita memiliki duplikat
df.duplicated().sum()

0

In [73]:
# Periksa ukuran dataset yang sekarang Anda miliki setelah manipulasi pertama yang Anda lakukan
df.head(20)

children  days_employed  dob_years            education  education_id  \
0        1.0    8437.673028       42.0    bachelor's degree             0   
1        1.0    4024.803754       36.0  secondary education             1   
2        0.0    5623.422610       33.0  secondary education             1   
3        3.0    4124.747207       32.0  secondary education             1   
4        0.0  340266.072047       53.0  secondary education             1   
5        0.0     926.185831       27.0    bachelor's degree             0   
6        0.0    2879.202052       43.0    bachelor's degree             0   
7        0.0     152.779569       50.0  secondary education             1   
8        2.0    6929.865299       35.0    bachelor's degree             0   
9        0.0    2188.756445       41.0  secondary education             1   
10       2.0    4171.483647       36.0    bachelor's degree             0   
11       0.0     792.701887       40.0  secondary education             1   
12       0.0            NaN       65.0  secondary education             1   
13       0.0    1846.641941       54.0         some college             2   
14       0.0    1844.956182       56.0    bachelor's degree             0   
15       1.0     972.364419       26.0  secondary education             1   
16       0.0    1719.934226       35.0  secondary education             1   
17       0.0    2369.999720       33.0    bachelor's degree             0   
18       0.0  400281.136913       53.0  secondary education             1   
19       0.0   10038.818549       48.0  secondary education             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate  
15     18691.345         construction of own property  
16   

[Jelaskan *dataset* baru Anda: jelaskan secara singkat apa saja perubahannya dan seberapa besar persentase perubahannya, jika ada.]

1. huruf pada kolom education sudah seragam
2. nilai -1 dan 20 pada kolom children sudah diganti menjadi nilai median
3. nilai XNA pada kolom gender sudah di drop karena tidak ada signifikansinya
4. duplikat sudah dihilangkan untuk menghindari pengulangan perhitungan atau double counting
5. nilai 0 pada kolom dob_years sudah diganti nilai median 
6. nilai minus pada kolom days_employed sudah menjadi positif untuk menghindari salah perhitungan

<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Oke, kesimpulan sementara sudah lengkap.
</div>

# Bekerja dengan nilai yang hilang

[Untuk mempercepat pekerjaan dengan sejumlah data, Anda mungkin ingin menggunakan *dictionary* untuk beberapa nilai yang memiliki ID. Jelaskan mengapa dan *dictionary* apakah yang akan Anda gunakan.]

In [74]:
# Temukan dictionary
#df.head()

In [75]:
education_dictionary=df[['education', 'education_id']].drop_duplicates().reset_index(drop=True)
education_dictionary

education  education_id
0    bachelor's degree             0
1  secondary education             1
2         some college             2
3    primary education             3
4      graduate degree             4

<div class="alert alert-block alert-danger">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Kamu perlu periksa lagi kenapa ada nilai `NaN` pada kolom `education`, kemungkinan besar karena kamu menggunakan `.head()` saat memperbaiki kolom `education` di atas.
</div>

<div class="alert alert-block alert-success">

<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Sip.
</div>

Saya hapus df.head() 

In [76]:
family_status_dictionary=df[['family_status', 'family_status_id']].drop_duplicates().reset_index(drop=True)
family_status_dictionary

family_status  family_status_id
0            married                 0
1  civil partnership                 1
2    widow / widower                 2
3           divorced                 3
4          unmarried                 4

<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Identifikasi dictionary untuk `family_status` sudah bagus.
</div>

### Memperbaiki nilai yang hilang di `total_income`

[Jelaskan secara singkat kolom dengan nilai yang hilang manakah yang perlu Anda tangani. Jelaskan bagaimana Anda akan memperbaikinya.]


[Mulailah dengan mengatasi total nilai pendapatan yang hilang. Buatlah kategori usia untuk nasabah. Buatlah kolom baru yang memuat kategori usia. Strategi ini dapat membantu untuk menghitung total nilai pendapatan.]


In [77]:
df['dob_years'].describe()

count    21452.000000
mean        43.469933
std         12.213723
min         19.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [78]:
df['dob_years'].value_counts()

42.0    696
35.0    616
40.0    607
41.0    605
34.0    601
38.0    597
33.0    581
39.0    572
31.0    559
36.0    554
44.0    545
29.0    544
30.0    537
48.0    536
37.0    536
50.0    513
43.0    512
32.0    509
49.0    508
28.0    503
45.0    496
27.0    493
52.0    484
56.0    483
47.0    477
54.0    476
46.0    472
53.0    459
57.0    456
58.0    454
51.0    446
55.0    443
59.0    443
26.0    408
60.0    374
25.0    357
61.0    354
62.0    348
63.0    269
24.0    263
64.0    260
23.0    252
65.0    193
22.0    183
66.0    182
67.0    167
21.0    111
68.0     99
69.0     85
70.0     65
71.0     56
20.0     51
72.0     33
19.0     14
73.0      8
74.0      6
75.0      1
Name: dob_years, dtype: int64

In [79]:
# Mari kita tulis sebuah fungsi untuk menghitung kategori usia

def age_group(age):
    try:
        if age<30:
            return '<30'
        if 31<=age<=40:
            return '31-40'
        if 41<=age<=50:
            return '41-50'
        if 51<=age<=60:
            return '51-60'
        if 61<=age<=70:
            return '61-70'
        else:
            return '70+'
    except:
        return 0

In [80]:
# Lakukan pengujian untuk melihat apakah fungsi Anda bekerja atau tidak
age_group(47)

'41-50'

In [81]:
# Buatlah kolom baru berdasarkan fungsi
df['age_group']=df['dob_years'].apply(age_group)
df

children  days_employed  dob_years            education  education_id  \
0           1.0    8437.673028       42.0    bachelor's degree             0   
1           1.0    4024.803754       36.0  secondary education             1   
2           0.0    5623.422610       33.0  secondary education             1   
3           3.0    4124.747207       32.0  secondary education             1   
4           0.0  340266.072047       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21447       1.0    4529.316663       43.0  secondary education             1   
21448       0.0  343937.404131       67.0  secondary education             1   
21449       1.0    2113.346888       38.0  secondary education             1   
21450       3.0    3112.481705       38.0  secondary education             1   
21451       2.0    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21447  civil partnership                 1      F    business     0   
21448            married                 0      F     retiree     0   
21449  civil partnership                 1      M    employee     1   
21450            married                 0      M    employee     1   
21451            married                 0      F    employee     0   

       total_income                  purpose age_group  
0         40620.102    purchase of the house     41-50  
1         17932.802             car purchase     31-40  
2         23341.752    purchase of the house     31-40  
3         42820.568  supplementary education     31-40  
4         25378.572        to have a wedding     51-60  
...             ...                      ...       ...  
21447     35966.698     housing transactions     41-50  
21448     24959.969        purchase of a car     61-70  
21449     14347.610                 property     31-40  
21450     39054.888        buying my own car     31-40  
21451     13127.587             to buy a car     31-40  

[21452 rows x 13 columns]

<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Kategorisasi `age_group` sudah bagus.
</div>

In [82]:
# Periksa bagaimana nilai di dalam kolom baru



[Pikirkan tentang faktor-faktor yang biasanya bergantung pada pendapatan. Pada akhirnya, Anda akan mengetahui apakah Anda harus menggunakan nilai rata-rata atau median untuk mengganti nilai yang hilang. Untuk membuat keputusan ini, Anda mungkin ingin melihat distribusi faktor-faktor yang Anda identifikasi sebagai dampak dari pendapatan seseorang.]

[Buatlah tabel yang hanya memuat data tanpa nilai yang hilang. Data ini akan digunakan untuk memperbaiki nilai yang hilang.]

In [83]:
# Buat tabel tanpa nilai yang hilang dan tampilkan beberapa barisnya untuk memastikan semuanya berjalan dengan baik
df_clean = df[df.notna()]
df_clean.head(20)

children  days_employed  dob_years            education  education_id  \
0        1.0    8437.673028       42.0    bachelor's degree             0   
1        1.0    4024.803754       36.0  secondary education             1   
2        0.0    5623.422610       33.0  secondary education             1   
3        3.0    4124.747207       32.0  secondary education             1   
4        0.0  340266.072047       53.0  secondary education             1   
5        0.0     926.185831       27.0    bachelor's degree             0   
6        0.0    2879.202052       43.0    bachelor's degree             0   
7        0.0     152.779569       50.0  secondary education             1   
8        2.0    6929.865299       35.0    bachelor's degree             0   
9        0.0    2188.756445       41.0  secondary education             1   
10       2.0    4171.483647       36.0    bachelor's degree             0   
11       0.0     792.701887       40.0  secondary education             1   
12       0.0            NaN       65.0  secondary education             1   
13       0.0    1846.641941       54.0         some college             2   
14       0.0    1844.956182       56.0    bachelor's degree             0   
15       1.0     972.364419       26.0  secondary education             1   
16       0.0    1719.934226       35.0  secondary education             1   
17       0.0    2369.999720       33.0    bachelor's degree             0   
18       0.0  400281.136913       53.0  secondary education             1   
19       0.0   10038.818549       48.0  secondary education             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   

    total_income                              purpose age_group  
0      40620.102                purchase of the house     41-50  
1      17932.802                         car purchase     31-40  
2      23341.752                purchase of the house     31-40  
3      42820.568              supplementary education     31-40  
4      25378.572                    to have a wedding     51-60  
5      40922.170                purchase of the house       <30  
6      38484.156                 housing transactions     41-50  
7      21731.829                            education     41-50  
8      15337.093                     having a wedding     31-40  
9      23108.150  purchase of the house for my family     41-50  
10     18230.959                      buy real estate     31-40  
11     12331.077           buy commercial real estate     31-40  
12           NaN                    to have a wedding     61-70  
13     20873.317                 

In [84]:
# Perhatikan nilai rata-rata untuk pendapatan berdasarkan faktor yang telah Anda identifikasi
total_income_mean=df.pivot_table(index='age_group', values='total_income', columns=['family_status','income_type'], aggfunc='mean')
total_income_mean

family_status civil partnership                                           \
income_type            business civil servant      employee entrepreneur   
age_group                                                                  
31-40              32226.488634  27971.002465  26188.970341          NaN   
41-50              34674.779759  25880.232662  26348.367906          NaN   
51-60              29943.399358  27588.084179  25260.372509          NaN   
61-70              25344.320333  18470.537143  25969.447000          NaN   
70+                35580.018565  20298.131700  24875.619915          NaN   
<30                28728.988670  25854.964684  23844.296331    79866.103   

family_status                               divorced                \
income_type         retiree unemployed      business civil servant   
age_group                                                            
31-40          17911.417273        NaN  35826.226443  28837.787048   
41-50          30144.495914  32435.602  29484.902216  30684.786345   
51-60          21557.702697        NaN  36228.896422  22480.691000   
61-70          22465.054700        NaN  53331.198625  33647.509333   
70+            18497.191750        NaN  39187.578778  23823.804500   
<30            12807.071000        NaN  32032.629348  22972.571143   

family_status                              ...    married     unmarried  \
income_type        employee       retiree  ... unemployed      business   
age_group                                  ...                            
31-40          26197.694901  19028.182000  ...   9593.119  31837.347112   
41-50          27108.936197  25302.179222  ...        NaN  34701.375513   
51-60          24404.466161  21479.611913  ...        NaN  32310.710853   
61-70          27539.349944  22901.630785  ...        NaN  32760.650187   
70+            27081.081778  23331.503000  ...        NaN  35137.774077   
<30            23016.427966  12066.799000  ...        NaN  28757.204337   

family_status                                                      \
income_type   civil servant      employee       retiree   student   
age_group                                                           
31-40          30505.072171  25237.303057  29331.419200       NaN   
41-50          30724.945242  26717.145691  26557.224632       NaN   
51-60          25432.284000  27399.417723  23810.634551       NaN   
61-70          35849.229500  37546.384842  20325.937656       NaN   
70+            25056.889375  25765.274911  18989.549444       NaN   
<30            25592.389711  23826.669745  11227.035000  15712.26   

family_status widow / widower                                            
income_type          business civil servant      employee       retiree  
age_group                                                                
31-40            29943.028615  27335.301333  25827.418333  34632.356000  
41-50            27085.293556  26605.605083  23521.346857  21514.467300  
51-60            27150.825333  22853.840421  23587.947322  20913.732735  
61-70            33531.175500  30257.731667  25622.147108  21649.627017  
70+                       NaN  24525.224000  25674.349500  17795.521000  
<30                       NaN           NaN  22915.450000           NaN  

[6 rows x 25 columns]

In [85]:
# Perhatikan nilai median untuk pendapatan berdasarkan faktor yang telah Anda identifikasi
total_income_median=df.pivot_table(index='age_group', values='total_income', aggfunc='median')
total_income_median

total_income
age_group              
31-40        24825.1865
41-50        24563.6500
51-60        22056.7710
61-70        19705.8550
70+          23083.2645
<30          22735.9110

[Ulangi perbandingan tersebut untuk beberapa faktor. Pastikan Anda mempertimbangkan berbagai aspek dan menjelaskan proses berpikir Anda.]



[Buatlah keputusan tentang karakteristik yang paling menentukan pendapatan dan apakah Anda akan menggunakan median atau rata-rata. Jelaskan mengapa Anda membuat keputusan ini]


In [86]:
#  Tulis fungsi yang akan kita gunakan untuk mengisi nilai yang hilang
df['total_income'] = df.groupby(['age_group'])['total_income'].transform(lambda x: x.fillna(x.median()))
        

In [87]:
# Memeriksa bagaimana nilai di dalam kolom baru
df.loc[(df['total_income'].isna())].shape

(0, 13)

In [88]:
# Terapkan fungsi tersebut ke setiap baris


In [89]:
# Periksa apakah kita mendapatkan kesalahan


[Jika Anda menemukan kesalahan dalam menyiapkan nilai untuk data yang hilang, kemungkinan ada sesuatu yang istimewa tentang data untuk kategori tersebut. Mari pikirkan - Anda mungkin ingin memperbaiki beberapa hal secara manual, jika terdapat cukup data untuk menemukan median/rata-rata.]


In [90]:
# Ganti nilai yang hilang jika terdapat kesalahan
df.isna().sum()

children               0
days_employed       2102
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
age_group              0
dtype: int64

[Setelah Anda selesai dengan `total_income`, periksa apakah jumlah total nilai di kolom ini sesuai dengan jumlah nilai di kolom lain.]

In [91]:
# Periksa jumlah entri di kolom
df.describe()


children  days_employed     dob_years  education_id  \
count  21452.000000   19350.000000  21452.000000  21452.000000   
mean       0.471331   66918.065141     43.469933      0.817033   
std        0.751093  139033.698578     12.213723      0.548639   
min        0.000000      24.141633     19.000000      0.000000   
25%        0.000000     926.990457     33.000000      1.000000   
50%        0.000000    2194.218768     42.000000      1.000000   
75%        1.000000    5538.423086     53.000000      1.000000   
max        5.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21452.000000  21452.000000   21452.000000  
mean           0.973895      0.081158   26446.496403  
std            1.421634      0.273084   15690.170869  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   17217.441750  
50%            0.000000      0.000000   23234.038000  
75%            1.000000      0.000000   31328.693750  
max            4.000000      1.000000  362496.645000

In [92]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21452 entries, 0 to 21451
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21452 non-null  float64
 1   days_employed     19350 non-null  float64
 2   dob_years         21452 non-null  float64
 3   education         21452 non-null  object 
 4   education_id      21452 non-null  int64  
 5   family_status     21452 non-null  object 
 6   family_status_id  21452 non-null  int64  
 7   gender            21452 non-null  object 
 8   income_type       21452 non-null  object 
 9   debt              21452 non-null  int64  
 10  total_income      21452 non-null  float64
 11  purpose           21452 non-null  object 
 12  age_group         21452 non-null  object 
dtypes: float64(4), int64(3), object(6)
memory usage: 2.1+ MB


<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Oke, mengisi nilai hilang pada `total_income` menggunakan `median` berdasarkan `age_group` bisa diterima.
</div>

###  Memperbaiki nilai di `days_employed`

[Pikirkan tentang parameter yang dapat membantu Anda memperbaiki nilai yang hilang di kolom ini. Pada akhirnya, Anda akan mengetahui apakah Anda harus menggunakan nilai rata-rata atau median untuk mengganti nilai yang hilang. Anda mungkin perlu melakukan penelitian yang sama dengan yang Anda lakukan saat memperbaiki data di kolom sebelumnya.]

In [93]:
# Distribusi median dari `days_employed` berdasarkan parameter yang Anda identifikasi

days_employed_median=df.pivot_table(index='age_group', values='days_employed', aggfunc='median')
days_employed_median

days_employed
age_group               
31-40        1630.193189
41-50        2199.733119
51-60        6481.039027
61-70      356081.013729
70+          1799.537390
<30           998.961907

In [94]:
# Distribusi rata-rata dari `days_employed` berdasarkan parameter yang Anda identifikasi
days_employed_mean=df.pivot_table(index='age_group', values='days_employed', aggfunc='mean')
days_employed_mean

days_employed
age_group               
31-40        4691.122800
41-50       17693.274606
51-60      153165.111212
61-70      288992.710281
70+         55910.432052
<30          2082.398127

[Tentukan apa yang akan Anda gunakan: rata-rata atau median. Jelaskan mengapa.]
median akan lebih masuk akal karena persebaran data (bobot nilai) tidak diketahui dengan jelas (akan lebih mudah jika dengan grafik visual).

In [95]:
# Mari tulis fungsi yang menghitung rata-rata atau median (tergantung keputusan Anda) berdasarkan parameter yang Anda identifikasi
df['days_employed'] = df.groupby(['age_group'])['days_employed'].transform(lambda x: x.fillna(x.median()))


In [96]:
df['days_employed'].describe()

count     21452.000000
mean      63985.511143
std      136316.866892
min          24.141633
25%         998.961907
50%        2199.733119
75%        5798.185967
max      401755.400475
Name: days_employed, dtype: float64

In [97]:
# Periksa apakah fungsi Anda dapat bekerja
df['days_employed'].head(50)


0       8437.673028
1       4024.803754
2       5623.422610
3       4124.747207
4     340266.072047
5        926.185831
6       2879.202052
7        152.779569
8       6929.865299
9       2188.756445
10      4171.483647
11       792.701887
12    356081.013729
13      1846.641941
14      1844.956182
15       972.364419
16      1719.934226
17      2369.999720
18    400281.136913
19     10038.818549
20      1311.604166
21       253.685166
22      1766.644138
23       272.981385
24    338551.952911
25    363548.489348
26      2199.733119
27       529.191635
28       717.274324
29    356081.013729
30    335581.668515
31      1682.083438
32      4649.910832
33      1548.637544
34      4488.067031
35    394021.072184
36       176.216688
37      6448.810860
38       597.881827
39       650.587796
40      1030.362606
41      2199.733119
42      1257.496190
43      4375.681384
44      1362.041728
45      1039.451741
46      2262.712304
47      2689.137274
48      3341.067886
49      1181.443228


In [98]:
# Terapkan fungsi ke income_type



In [99]:
# Periksa apakah fungsi Anda bekerja

df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_group           0
dtype: int64

In [100]:
# Ganti nilai yang hilang



[Setelah Anda selesai dengan `total_income`, periksa apakah jumlah total nilai di kolom ini sesuai dengan jumlah nilai di kolom lain.]

In [101]:
# Periksa entri di semua kolom - pastikan kita memperbaiki semua nilai yang hilang
df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_group           0
dtype: int64

In [102]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21452 entries, 0 to 21451
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21452 non-null  float64
 1   days_employed     21452 non-null  float64
 2   dob_years         21452 non-null  float64
 3   education         21452 non-null  object 
 4   education_id      21452 non-null  int64  
 5   family_status     21452 non-null  object 
 6   family_status_id  21452 non-null  int64  
 7   gender            21452 non-null  object 
 8   income_type       21452 non-null  object 
 9   debt              21452 non-null  int64  
 10  total_income      21452 non-null  float64
 11  purpose           21452 non-null  object 
 12  age_group         21452 non-null  object 
dtypes: float64(4), int64(3), object(6)
memory usage: 2.1+ MB


<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Mengisi nilai hilang pada `days_employed` menggunakan `median` berdasarkan `age_group` bisa diterima.
</div>

## Pengkategorian Data

[Untuk menjawab pertanyaan dan menguji hipotesis, Anda akan bekerja dengan data yang telah dikategorikan. Lihatlah pertanyaan-pertanyaan yang diajukan kepada Anda dan yang harus Anda jawab. Pikirkan tentang data mana yang perlu dikategorikan untuk menjawab pertanyaan-pertanyaan ini. Di bawah ini, Anda akan menemukan templat yang bisa Anda gunakan untuk mengkategorikan data. Proses pertama mencakup data teks; yang kedua membahas data numerik yang perlu dikategorikan. Anda dapat menggunakan kedua petunjuk yang disarankan atau tidak sama sekali - semuanya terserah Anda.]

[Terlepas dari keputusan Anda untuk mengatasi pengkategorian, pastikan bahwa Anda secara lugas memberikan penjelasan tentang mengapa Anda membuat keputusan tersebut. Ingat: ini adalah pekerjaan Anda dan hanya Anda yang berhak membuat segala keputusan.]


In [103]:
# Tampilkan nilai data yang Anda pilih untuk pengkategorian

df['purpose'].value_counts()


wedding ceremony                            791
having a wedding                            767
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
housing                                     646
purchase of the house                       646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
purchase of my own house                    620
buy real estate                             620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

[Mari kita memeriksa nilai unik]

In [104]:
# Periksa nilai unik
df['purpose'].sort_values().unique()

array(['building a property', 'building a real estate',
       'buy commercial real estate', 'buy real estate',
       'buy residential real estate', 'buying a second-hand car',
       'buying my own car', 'buying property for renting out', 'car',
       'car purchase', 'cars', 'construction of own property',
       'education', 'getting an education', 'getting higher education',
       'going to university', 'having a wedding', 'housing',
       'housing renovation', 'housing transactions', 'profile education',
       'property', 'purchase of a car', 'purchase of my own house',
       'purchase of the house', 'purchase of the house for my family',
       'real estate transactions', 'second-hand car purchase',
       'supplementary education', 'to become educated', 'to buy a car',
       'to get a supplementary education', 'to have a wedding',
       'to own a car', 'transactions with commercial real estate',
       'transactions with my real estate', 'university education',
       'we

[Kelompok utama apakah yang dapat Anda identifikasi berdasarkan nilai uniknya?]

[Kita akan mengkategorikan data kita berdasarkan topik ini.]


In [105]:
# Mari kita tulis sebuah fungsi untuk mengategorikan data berdasarkan topik umum
def purpose_common(purpose):
    
    if 'property' in purpose:
        return 'property'
    elif 'estate' in purpose:
        return 'property'
    elif 'hous' in purpose:
        return 'property'
    elif 'car' in purpose:
        return 'car'
    elif 'educ' in purpose:
        return 'education'
    elif 'univ' in purpose:
        return 'education'
    elif 'wedd' in purpose:
        return 'wedding'

In [106]:
# Buat kolom yang memuat kategori dan hitung nilainya
df['general_purpose'] = df['purpose'].apply(purpose_common)
print(df['general_purpose'].value_counts())
print()
df['general_purpose'].count()


property     10810
car           4306
education     4013
wedding       2323
Name: general_purpose, dtype: int64



21452

<div class="alert alert-block alert-success">

<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Sip, untuk kategorisasi purpose sudah bagus.
</div>

[Jika Anda memutuskan untuk mengategorikan data numerik, Anda juga harus membuat kategori untuk data tersebut.]

In [107]:
# Lihat semua data numerik di kolom yang Anda pilih untuk pengkategorian


In [108]:
# Dapatkan kesimpulan statistik untuk kolomnya
# telah dikategorikan kedalam 4 kategori: property, car, education dan wedding. 


[Tentukan rentang apa yang akan Anda gunakan untuk pengelompokan dan jelaskan alasannya.]

In [109]:
# Buat fungsi yang melakukan pengkategorian menjadi kelompok numerik yang berbeda berdasarkan rentang



In [110]:
# Buat kolom yang memuat kategori
df

children  days_employed  dob_years            education  education_id  \
0           1.0    8437.673028       42.0    bachelor's degree             0   
1           1.0    4024.803754       36.0  secondary education             1   
2           0.0    5623.422610       33.0  secondary education             1   
3           3.0    4124.747207       32.0  secondary education             1   
4           0.0  340266.072047       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21447       1.0    4529.316663       43.0  secondary education             1   
21448       0.0  343937.404131       67.0  secondary education             1   
21449       1.0    2113.346888       38.0  secondary education             1   
21450       3.0    3112.481705       38.0  secondary education             1   
21451       2.0    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21447  civil partnership                 1      F    business     0   
21448            married                 0      F     retiree     0   
21449  civil partnership                 1      M    employee     1   
21450            married                 0      M    employee     1   
21451            married                 0      F    employee     0   

       total_income                  purpose age_group general_purpose  
0         40620.102    purchase of the house     41-50        property  
1         17932.802             car purchase     31-40             car  
2         23341.752    purchase of the house     31-40        property  
3         42820.568  supplementary education     31-40       education  
4         25378.572        to have a wedding     51-60         wedding  
...             ...                      ...       ...             ...  
21447     35966.698     housing transactions     41-50        property  
21448     24959.969        purchase of a car     61-70             car  
21449     14347.610                 property     31-40        property  
21450     39054.888        buying my own car     31-40             car  
21451     13127.587             to buy a car     31-40             car  

[21452 rows x 14 columns]

In [111]:
def income_class(total_income):
 
    if total_income <= 32000:
        return 'poor'
    elif total_income <= 53000:
        return 'lower middle class'
    elif total_income <= 106000:
        return 'middle class'
    elif total_income <= 373000:
        return 'upper middle class'
    else:
        return 'rich'

Saya mengkategorikan total_income ke dalam beberapa kelas kategori berdasarkan klasifikasi tingkat ekonomi yaitu poor, lower middle class, middle class, upper middle class dan rich. Hal ini ditujukan untuk menggolongkan kategori saja sehingga memudahkan dalamn proses pengambilan keputusan.

<div class="alert alert-block alert-warning">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Kamu perlu menjelaskan ya kenapa angka-angka di atas dipilih sebagai batasan masing-masing kategori
</div>

<div class="alert alert-block alert-success">

<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Oke.
</div>

In [112]:
df['economic_class'] = df['total_income'].apply(income_class)

In [113]:
df['economic_class'].value_counts()



poor                  16387
lower middle class     3985
middle class           1000
upper middle class       80
Name: economic_class, dtype: int64

In [114]:
# Hitung setiap nilai kategori untuk melihat pendistribusiannya
df

children  days_employed  dob_years            education  education_id  \
0           1.0    8437.673028       42.0    bachelor's degree             0   
1           1.0    4024.803754       36.0  secondary education             1   
2           0.0    5623.422610       33.0  secondary education             1   
3           3.0    4124.747207       32.0  secondary education             1   
4           0.0  340266.072047       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21447       1.0    4529.316663       43.0  secondary education             1   
21448       0.0  343937.404131       67.0  secondary education             1   
21449       1.0    2113.346888       38.0  secondary education             1   
21450       3.0    3112.481705       38.0  secondary education             1   
21451       2.0    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21447  civil partnership                 1      F    business     0   
21448            married                 0      F     retiree     0   
21449  civil partnership                 1      M    employee     1   
21450            married                 0      M    employee     1   
21451            married                 0      F    employee     0   

       total_income                  purpose age_group general_purpose  \
0         40620.102    purchase of the house     41-50        property   
1         17932.802             car purchase     31-40             car   
2         23341.752    purchase of the house     31-40        property   
3         42820.568  supplementary education     31-40       education   
4         25378.572        to have a wedding     51-60         wedding   
...             ...                      ...       ...             ...   
21447     35966.698     housing transactions     41-50        property   
21448     24959.969        purchase of a car     61-70             car   
21449     14347.610                 property     31-40        property   
21450     39054.888        buying my own car     31-40             car   
21451     13127.587             to buy a car     31-40             car   

           economic_class  
0      lower middle class  
1                    poor  
2                    poor  
3      lower middle class  
4                    poor  
...                   ...  
21447  lower middle class  
21448                poor  
21449                poor  
21450  lower middle class  
21451                poor  

[21452 rows x 15 columns]

## Memeriksa hipotesis


**Apakah terdapat korelasi antara memiliki anak dengan probabilitas melakukan gagal bayar pinjaman?**

In [115]:
# Periksa data anak dan data gagal bayar pinjaman


# Hitung persentase gagal bayar berdasarkan jumlah anak

pivot_table_children = df.pivot_table(index='children', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_children

debt            0       1
children                 
0.0       13140.0  1072.0
1.0        4364.0   444.0
2.0        1858.0   194.0
3.0         303.0    27.0
4.0          37.0     4.0
5.0           9.0     NaN

In [116]:
df['debt'].value_counts()

0    19711
1     1741
Name: debt, dtype: int64

In [117]:
pivot_table_children['percentage'] = pivot_table_children[1] / (pivot_table_children[1] + pivot_table_children[0]) * 100
pivot_table_children

debt            0       1  percentage
children                             
0.0       13140.0  1072.0    7.542921
1.0        4364.0   444.0    9.234609
2.0        1858.0   194.0    9.454191
3.0         303.0    27.0    8.181818
4.0          37.0     4.0    9.756098
5.0           9.0     NaN         NaN

**Kesimpulan**

[Tulis kesimpulan Anda berdasarkan manipulasi dan pengamatan yang Anda lakukan.]
1.calon debitur yang tidak punya anak, akan memiliki peluang 'hanya' 7.5% untuk gagal bayar / terbeban hutang
2.calon debitur yang memiliki 1 sd 4 anak akan memiliki peluang gagal bayar 8-9.7% 

<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Sudah bagus, bisa kamu simpulkan juga bahwa jika calon debitur memiliki anak, peluang gagal bayarnya cenderung lebih tinggi daripada yang belum memiliki anak.
</div>

**Apakah terdapat korelasi antara status keluarga dengan probabilitas melakukan gagal bayar pinjaman?**

In [118]:
# Periksa data status keluarga dan data gagal bayar pinjaman



# Hitung persentase gagal bayar berdasarkan status keluarga

pivot_table_family_status = df.pivot_table(index='family_status', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_family_status



debt                   0    1
family_status                
civil partnership   3761  388
divorced            1110   85
married            11408  931
unmarried           2536  274
widow / widower      896   63

In [119]:
pivot_table_family_status['percentage'] = pivot_table_family_status[1] / (pivot_table_family_status[1] + pivot_table_family_status[0]) * 100
pivot_table_family_status

debt                   0    1  percentage
family_status                            
civil partnership   3761  388    9.351651
divorced            1110   85    7.112971
married            11408  931    7.545182
unmarried           2536  274    9.750890
widow / widower      896   63    6.569343

**Kesimpulan**

[Tulis kesimpulan Anda berdasarkan manipulasi dan pengamatan Anda.]
1. calon debitur dengan kondisi unmarried dan civil partnership (pasangan) akan memiliki kemungkinan gagal bayar terbesar (9.8% dan 9.4%)
2. sedangakan calon debitur dengan kondisi widow/widower 'hanya' memiliki peluang gagal bayar terendah (7%)



<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Sip.
</div>

**Apakah terdapat korelasi antara tingkat pendapatan dengan probabilitas melakukan gagal bayar pinjaman?**

In [120]:
# Periksa data tingkat pendapatan dan data gagal bayar pinjaman



# Hitung persentase gagal bayar berdasarkan tingkat pendapatan

pivot_table_economic_class = df.pivot_table(index='economic_class', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_economic_class


debt                    0     1
economic_class                 
lower middle class   3704   281
middle class          928    72
poor                15004  1383
upper middle class     75     5

In [121]:
pivot_table_economic_class['percentage'] = pivot_table_economic_class[1] / (pivot_table_economic_class[1] + pivot_table_economic_class[0]) * 100
pivot_table_economic_class


debt                    0     1  percentage
economic_class                             
lower middle class   3704   281    7.051443
middle class          928    72    7.200000
poor                15004  1383    8.439617
upper middle class     75     5    6.250000

**Kesimpulan**

[Tulis kesimpulan Anda berdasarkan manipulasi dan pengamatan Anda.]
1. semakin mengarah pada kategori 'poor' maka peluang gagal bayar semakin besar

<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Oke.
</div>

**Bagaimana tujuan kredit memengaruhi persentase gagal bayar?**

In [122]:
# Periksa persentase tingkat gagal bayar untuk setiap tujuan kredit dan lakukan penganalisisan

pivot_table_general_purpose = df.pivot_table(index='general_purpose', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_general_purpose['percentage'] = pivot_table_general_purpose[1] / (pivot_table_general_purpose[1] + pivot_table_general_purpose[0]) * 100
pivot_table_general_purpose


debt                 0    1  percentage
general_purpose                        
car               3903  403    9.359034
education         3643  370    9.220035
property         10028  782    7.234043
wedding           2137  186    8.006888

**Kesimpulan**

[Tulis kesimpulan Anda berdasarkan manipulasi dan pengamatan yang Anda lakukan.]
1. kemungkinan tertinggi gagal bayar berdasarkan data diatas maka dapat disimpulkan bahwa calon debitur dengan tujuan untuk membeli atau menyewa car dan atau membiayai education (>9%)
2. sedangkan untuk yang bertujuan membeli atau menyewa property memiliki peluang gagal bayar terendah. 




<div class="alert alert-block alert-success">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Bagus sekali.
</div>

# Kesimpulan umum 

[Tuliskan kesimpulan Anda di bagian akhir ini. Pastikan Anda memasukkan semua kesimpulan penting yang telah Anda buat berkaitan dengan cara Anda memproses dan menganalisis data. Kesimpulan tersebut harus membahas nilai yang hilang, duplikat, dan kemungkinan alasan serta solusi untuk data bermasalah yang harus Anda tangani.]

[Tuliskan juga kesimpulan Anda mengenai pertanyaan-pertanyaan yang diajukan di sini.]


<div class="alert alert-block alert-danger">

<b>Reviewer's comment v1</b> <a class="tocSkip"></a>

Jangan lupa tuliskan apa saja yang telah kamu lakukan pada notebook ini di bagian kesimpulan akhir ya. Tambahkan juga dengan hasil analisis kamu tentang korelasi setiap kolom pada data terhadap kemungkinan gagal bayar, sehingga pembaca bisa mendapatkan *insight* yang berguna untuk bisnis melalui hasil proyek kamu.
</div>

<div class="alert alert-block alert-success">

<b>Reviewer's comment v2</b> <a class="tocSkip"></a>

Bagus sekali, kesimpulan sudah lengkap.
</div>

Kesimpulan:
1. Terdapat nilai yang hilang pada dataset ini ada secara acak tanpa pola. Dalam mengatasi data hilang ini dilakukan dengan pengisian data dari nilai median data yang tersedia.
2. Data duplikat pada dataset ini diperbaiki dengan mengubah cara penulisan menjadi huruf kecil.
3. Terdapat pula nilai data yang aneh pada kolom 'children' yaitu memiliki anak berjumlah -1 dan 20. Data tersebut diperbaiki dengan mengubah nilai negatif menjadi positif dan tidak menghapus nilai data yang ada karena kurangnya informasi yang meyakinkan untuk menghapus nilai data. 
4. Pada kolom 'dob_years', terdapat nilai data yang bermalah yaitu bernilai 0. Data tersebut diperbaiki dengan mengganti nilai tersebut dengan median yaitu 42. 
5. Data baru yang telah diperbaiki diperiksa kembali duplikatnya. Dilakukan proses drop duplikat.
6. Data dictionary dilakukan drop duplikat karena merupakan salah satu cara mudah untuk memperbaiki data.
7. Nilai yang hilang pada kolom 'total_income' diperbaiki dengan membuat kategori usia untuk nasabah. Lalu diisi dengan nilai median dari 'total_income'.
8. Nilai yang hilang pada kolom 'days_employed' diperbaiki dengan mengisi nilai kosong dengan data median yang tersedia.
9. Kategori data untuk mempermudah pengambil keputusan dan pemeriksaan hipotesis.
10. Dari hasil pemeriksaan data hipotesis, diketahui bahwa probabilitas nasabah dalam memiliki anak, status keluarga, total pendapatan, dan tujuan pembelian (dapat dikatakantidak mempengaruhi gagal bayar nasabah karena diketahui pada data bahwa 90% ke atas nasabah berhasil membayar pinjaman.
